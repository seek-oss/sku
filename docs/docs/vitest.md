# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
