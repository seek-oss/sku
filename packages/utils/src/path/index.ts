export { toPosixPath } from './toPosixPath.js';
