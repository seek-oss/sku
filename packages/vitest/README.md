[![npm](https://img.shields.io/npm/v/@sku-lib/vitest.svg?style=flat-square)](https://www.npmjs.com/package/@sku-lib/vitest)

# vitest runner for `sku test`
