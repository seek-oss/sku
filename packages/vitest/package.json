{
  "name": "@sku-lib/vitest",
  "version": "0.2.2",
  "description": "vitest runner for sku test",
  "types": "./dist/index.d.ts",
  "main": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "engines": {
    "node": ">=22.19.0"
  },
  "files": [
    "dist"
  ],
  "type": "module",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "lint:tsc": "tsc --noEmit"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/seek-oss/sku.git",
    "directory": "packages/vite"
  },
  "author": "SEEK",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/seek-oss/sku/issues"
  },
  "homepage": "https://github.com/seek-oss/sku#readme",
  "devDependencies": {
    "typescript-transform-paths": "^3.5.2",
    "vitest": "catalog:"
  },
  "dependencies": {
    "@vanilla-extract/vite-plugin": "catalog:",
    "vite-tsconfig-paths": "^5.1.4"
  },
  "peerDependencies": {
    "vitest": "^3.0.0"
  }
}
