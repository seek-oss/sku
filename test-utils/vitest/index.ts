export { getAppSnapshot } from './appSnapshot.ts';
