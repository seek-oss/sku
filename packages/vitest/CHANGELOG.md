# @sku-lib/vitest

## 0.2.1

### Patch Changes

- Fix `watch` mode ([#1393](https://github.com/seek-oss/sku/pull/1393))

## 0.2.0

### Minor Changes

- Disable Vitest globals ([#1387](https://github.com/seek-oss/sku/pull/1387))

  `vitest` is now a peer dependency. As such, `vitest` APIs must now be imported explicitly

  **EXAMPLE USAGE**:

  ```ts
  // example.test.ts
  import { describe, it, expect } from 'vitest';

  describe("test suite", () => {
    it("should be true" () => {
      expect(true).toBe(true);
    });
  });
  ```

### Patch Changes

- `deps`: Update `vite-plugin-cjs-interop` from `^2.2.0` to `^2.3.0` ([#1386](https://github.com/seek-oss/sku/pull/1386))

- Added missing `compilePackages` and `cjsInteropDependencies` support ([#1381](https://github.com/seek-oss/sku/pull/1381))

- Fixed a bug causing `sku test` to hang when `vitest` is executed in `run` mode ([#1387](https://github.com/seek-oss/sku/pull/1387))

## 0.1.1

### Patch Changes

- Experimental vite bundler now requires vite v7. ([#1346](https://github.com/seek-oss/sku/pull/1346))

  This is a breaking peer dependency change for projects that use the experimental vite bundler. Since this is an experimental feature, we are not bumping the major version.

- - `vitest`: Honour watch mode when running `sku test` ([#1325](https://github.com/seek-oss/sku/pull/1325))

    Previously it was not possible to run Vitest in watch mode via `sku test`. `sku test` should now behave similarly to [the `vitest` CLI][vitest cli], so `sku test`, `sku test watch` and `sku test -w/--watch` will all trigger watch mode.

    To run tests once and exit, you can execute `sku test run`, `sku test --run` or `sku test --no-watch` .

    [vitest cli]: https://vitest.dev/guide/cli

  - `deps` Bump peer dependencies to support React 19

## 0.1.0

### Minor Changes

- Initial release. ([#1273](https://github.com/seek-oss/sku/pull/1273))
