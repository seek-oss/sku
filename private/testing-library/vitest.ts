// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
