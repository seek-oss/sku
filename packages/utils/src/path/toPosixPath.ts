import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
