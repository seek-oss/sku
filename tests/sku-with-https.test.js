import { describe, beforeAll, afterAll, it } from 'vitest';
import { getAppSnapshot } from '@sku-private/vitest-utils';
import path from 'node:path';
import fs from 'node:fs/promises';
import {
  getPort,
  runSkuScriptInDir,
  waitForUrls,
  createCancelSignal,
} from '@sku-private/test-utils';

import skuServerConfig from '@sku-fixtures/sku-with-https/sku-server.config.mjs';

import { createRequire } from 'node:module';

const { serverPort } = skuServerConfig;

const require = createRequire(import.meta.url);

const appDir = path.dirname(
  require.resolve('@sku-fixtures/sku-with-https/sku.config.mjs'),
);

describe.sequential('sku-with-https', () => {
  describe.concurrent.each(['vite', 'webpack'])(
    'bundler: %s',
    async (bundler) => {
      const port = await getPort();
      const args = ['--strict-port', `--port=${port}`];
      if (bundler === 'vite')
        args.push(
          ...['--experimental-bundler', '--config', 'sku.config.vite.mjs'],
        );
      describe('start', () => {
        const { cancel, signal } = createCancelSignal();
        const url = `https://localhost:${port}`;

        beforeAll(async () => {
          runSkuScriptInDir('start', appDir, args, {
            signal,
          });
          await waitForUrls(url, `${url}/test-middleware`);
        });

        afterAll(async () => {
          cancel();
        });

        it('should start a development server', async ({ expect }) => {
          const snapshot = await getAppSnapshot({ url, expect });
          expect(snapshot).toMatchSnapshot();
        });
        it('should support the supplied middleware', async ({ expect }) => {
          const snapshot = await getAppSnapshot({
            url: `${url}/test-middleware`,
            expect,
          });
          expect(snapshot).toMatchSnapshot();
        });
      });
    },
  );

  describe('start-ssr', () => {
    const url = `https://localhost:${serverPort}`;
    const { cancel, signal } = createCancelSignal();

    beforeAll(async () => {
      runSkuScriptInDir(
        'start-ssr',
        appDir,
        ['--config=sku-server.config.mjs'],
        { signal },
      );
      await waitForUrls(url, `${url}/test-middleware`);
    });

    afterAll(async () => {
      cancel();
    });

    it('should support the supplied middleware', async ({ expect }) => {
      const snapshot = await getAppSnapshot({
        url: `${url}/test-middleware`,
        expect,
      });
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('serve', async () => {
    const port = await getPort();
    const url = `https://localhost:${port}`;
    const { cancel, signal } = createCancelSignal();

    beforeAll(async () => {
      await runSkuScriptInDir('build', appDir);
      runSkuScriptInDir('serve', appDir, ['--strict-port', `--port=${port}`], {
        signal,
      });
      await waitForUrls(url, `${url}/test-middleware`);
    });

    afterAll(async () => {
      cancel();
    });

    it('should start a development server', async ({ expect }) => {
      const snapshot = await getAppSnapshot({ url, expect });
      expect(snapshot).toMatchSnapshot();
    });

    it('should support the supplied middleware', async ({ expect }) => {
      const snapshot = await getAppSnapshot({
        url: `${url}/test-middleware`,
        expect,
      });
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('.gitignore', () => {
    it('should add the .ssl directory to .gitignore', async ({ expect }) => {
      const ignoreContents = await fs.readFile(
        path.join(appDir, '.gitignore'),
        'utf-8',
      );
      expect(ignoreContents.split('\n')).toContain(`.ssl`);
    });
  });
});
