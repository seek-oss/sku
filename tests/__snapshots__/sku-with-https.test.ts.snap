// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sku-with-https > bundler: vite > start > should start a development server 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://some-cdn.com 'sha256-Z2/iFzh9VMlVkEOar1f/oSHWwQk3ve1qk/C2WdsC4Xk=';"
    >
    <script type="module">
      import { injectIntoGlobalHook } from "/@react-refresh";
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;
    </script>
    <script
      type="module"
      src="/@vite/client"
    >
    </script>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=0.js"
    >
    </script>
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=1.js"
    >
    </script>
  </head>
  <body>
    <div id="app">
      <link
        rel="preload"
        as="image"
        href="/src/logo.png"
      >
      <div class="styles_root__2vdre50">
        <div class="styles_nested__2vdre51">
          Initial
        </div>
        <img src="/src/logo.png">
      </div>
    </div>
    <script
      type="module"
      src="{cwd}/packages/sku/dist/services/vite/entries/vite-client.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -26,14 +26,25 @@
     >
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=0.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=0.js"
     >
     </script>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=1.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=1.js"
     >
     </script>
+    <style
+      type="text/css"
+      data-vite-dev-id="{cwd}/fixtures/sku-with-https/src/styles.css.js.vanilla.css"
+    >
+      .styles_root__2vdre50 {
+  display: flex;
+}
+.styles_root__2vdre50 .styles_nested__2vdre51 {
+  font-size: 32px;
+}
+    </style>
   </head>
   <body>
     <div id="app">
@@ -44,7 +55,7 @@
       >
       <div class="styles_root__2vdre50">
         <div class="styles_nested__2vdre51">
-          Initial
+          Client
         </div>
         <img src="/src/logo.png">
       </div>
`;

exports[`sku-with-https > bundler: vite > start > should support the supplied middleware 1`] = `
"OK"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -1,1 +1,13 @@
-OK
\\ No newline at end of file
+<html>
+  <head>
+    <meta
+      name="color-scheme"
+      content="light dark"
+    >
+  </head>
+  <body>
+    <pre style="word-wrap: break-word; white-space: pre-wrap;">
+      OK
+    </pre>
+  </body>
+</html>
\\ No newline at end of file
`;

exports[`sku-with-https > bundler: webpack > start > should start a development server 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://some-cdn.com 'unsafe-eval';"
    >
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="/main.css"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/main.js"
    >
  </head>
  <body>
    <div id="app">
      <link
        rel="preload"
        as="image"
        href="/a85d2645cc4d80a20b5b.png"
      >
      <div class="styles_root__2vdre50">
        <div class="styles_nested__2vdre51">
          Initial
        </div>
        <img src="/a85d2645cc4d80a20b5b.png">
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/main.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -46,7 +46,7 @@
       >
       <div class="styles_root__2vdre50">
         <div class="styles_nested__2vdre51">
-          Initial
+          Client
         </div>
         <img src="/a85d2645cc4d80a20b5b.png">
       </div>
`;

exports[`sku-with-https > bundler: webpack > start > should support the supplied middleware 1`] = `
"OK"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -1,1 +1,7 @@
-OK
\\ No newline at end of file
+<html>
+  <head>
+  </head>
+  <body>
+    OK
+  </body>
+</html>
\\ No newline at end of file
`;

exports[`sku-with-https > serve > should start a development server 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://some-cdn.com;"
    >
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="/main-20b517520995ff660440.css"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/main.js"
    >
  </head>
  <body>
    <div id="app">
      <link
        rel="preload"
        as="image"
        href="/a85d2645cc4d80a20b5b.png"
      >
      <div class="_2vdre50">
        <div class="_2vdre51">
          Initial
        </div>
        <img src="/a85d2645cc4d80a20b5b.png">
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/main.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -46,7 +46,7 @@
       >
       <div class="_2vdre50">
         <div class="_2vdre51">
-          Initial
+          Client
         </div>
         <img src="/a85d2645cc4d80a20b5b.png">
       </div>
`;

exports[`sku-with-https > serve > should support the supplied middleware 1`] = `
"OK"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -1,1 +1,7 @@
-OK
\\ No newline at end of file
+<html>
+  <head>
+  </head>
+  <body>
+    OK
+  </body>
+</html>
\\ No newline at end of file
`;

exports[`sku-with-https > start-ssr > should support the supplied middleware 1`] = `
"OK"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -1,1 +1,7 @@
-OK
\\ No newline at end of file
+<html>
+  <head>
+  </head>
+  <body>
+    OK
+  </body>
+</html>
\\ No newline at end of file
`;
